2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+Install-Module Az.KeyVault
+
+# In this example, we explicitly register this extension
+Register-SecretsVaultExtension -Name Azure -Cmdlet Get-AzKeyVaultSecret `
+  -Module Az.KeyVault -Version 1.0.0
+
+# When using remote vaults, expectation is that secrets are
+# stored using their tooling, so this module is only for retrieving secrets
+# from remote vaults
+$azDevOpsToken = Get-Secret -Name AzDevOps
+Invoke-RestMethod https://dev.azure.com/… -Credential $azDevOpsToken -Authentication Basic
+```
+
+Registering and using local secret:
+
+```powershell
+# For local vault, we can register custom secrets
+# In this example, we store a PSCredential object
+Register-Secret -Secret $cred -Name MyCreds
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
+passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
+and the module version to `-ModuleVersion`.
+This can be done as part of importing the module or explicitly by the user.
+
+`Register-SecretsVaultExtension` has a `-Default` switch to register the
+default local vault (if applicable, see above) if it has been unregistered.
+
+`Register-SecretsVaultExtension` takes a `-AccessToken` or `-Credential` parameter
+for authenticating with the remote vault.
+If neither are supplied, then the current user context is used and relies on
+the extension cmdlet to handle authentication.
+
+`Unregister-SecretsVaultExtension` should be used to remove any existing
+registration or when updating a registration to a new version.
+
+`Get-SecretsVaultExtension` will enumerate registered extensions returning
+the module name, module version, and cmdlet used.
+
+### Storing Secrets
+
+The `Register-Secret` cmdlet is used to store a secret.
+The `-Name` must be unique across all secrets regardless if local or remote.
```

Comment:
Have you considered that `-Name` should perhaps be `-Path`?

E.g. HashiCorp Vault, KeePass, SecretServer, etc. use folders and thus, paths to identify a secret

Reply:
+1 for Path.

Reply:
So `-Path` can be passed as part of the hashtable (new thing being added) at `Register-SecretsVault` (I think we should drop the `Extension` suffix).  So you can register different HashiCorp Vaults, for example, that map to different paths.  `Add-Secret` would just take the `-Vault` name that maps to the registration.